Reduce the amount of per-table metrics

amnonh opened the pull request:
This series is the first step in the effort to reduce the number of metrics reported by Scylla.
The series focuses on the per-table metrics.

The combination of histograms, per-tables, and per shard makes the number of metrics in a cluster explode.
The following series uses multiple tools to reduce the number of metrics.
1. Multiple metrics should only be reported for the user tables and the condition that checked it was not updated when more non-user keyspaces were added.
2. Second, instead of a histogram, per table, per shard, it will report a summary per table, per shard, and a single histogram per node.
3. Histograms, summaries, and counters will be reported only if they are used (for example, the cas-related metrics will not be reported for tables that are not using cas).

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/1392/

avikivity commented on utils/histogram.hh:
```diff
@@ -118,9 +118,13 @@ public:
-    bool should_sample() const {
-        return total == 0 || (started & sample_mask);
+    bool should_sample() const noexcept {
+        return total == 0 || ((started & sample_mask) == sample_mask);
```
Maybe deserves a bug # and backports. What's the impact?

This sample_mask is very unintuitive, we should replace it (but ok after the series).

amnonh replied:
I think @tgrabiec found it years ago, I couldn't find an issue, I'm planning to deprecate this class alltogether but currently, this bug prevent disable the sampling.

Currently, what the code does is sample half of the points instead of according to the mask (1 out of 1^n).

tgrabiec replied:
https://github.com/scylladb/scylla/issues/2747

avikivity commented on utils/histogram.hh:
```diff
@@ -264,17 +265,101 @@ public:
+ * On each time tick, a summary is calculated and stored locally and the
+ * estimated histogram is updated.
+ */
+class timed_rate_moving_average_with_summary : public timed_rate_moving_average {
```
I think composition is better than inheritance here. It avoids using virtual functions, and I'm guessing there will be no problem calling timed_rate_moving_average_with_summary.get_timed_rate_moving_average().

avikivity replied:
Right.

I think it makes sense to move the timer out and leave this as a pure data+math class with no mechanics.

A (very small) benefit is that it allows consolidating timers, a single timer can update many moving averages. But it's good to have such classes focus on the math and not setup/teardown etc.

avikivity commented on utils/histogram_metrics_helper.cc:
```diff
@@ -0,0 +1,35 @@
+/*
+ * Copyright (C) 2020-present ScyllaDB
```
2022

amnonh replied:
I'll update

avikivity commented on utils/histogram_metrics_helper.cc:
```diff
@@ -0,0 +1,35 @@
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with Scylla.  If not, see <http://www.gnu.org/licenses/>.
+ */
```
SPDX

amnonh replied:
ok

amnonh commented on utils/histogram_metrics_helper.cc:
```diff
@@ -0,0 +1,35 @@
+ * Copyright (C) 2020-present ScyllaDB
+ *
+ * Modified by ScyllaDB
+ */
```
it's not, I'll remove

amnonh commented:
@avikivity I've updated the license header to the newer format. About ```class timed_rate_moving_average_with_summary``` I'm not sure which direction you prefer, should I break the timer functionality out of timed_rate_moving_average and split the code accordingly?
Do you have another direction in mind?

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/1430/

avikivity commented:
> @avikivity I've updated the license header to the newer format. About `class timed_rate_moving_average_with_summary` I'm not sure which direction you prefer, should I break the timer functionality out of timed_rate_moving_average and split the code accordingly? Do you have another direction in mind?

I did like moving the timer out.

amnonh commented:
V2 breaks the timer functionality into a data class and a timer class.

To make the patches more readable, this is done in two stages, on patch breaks ```timed_rate_moving_average``` into a data class and a timer, and another patch adds the summary calculation class.

Breaking ```timed_rate_moving_average``` into two classes required some code change in places that metrics used internal member, now it uses a method.

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/1464/

amnonh commented:
@avikivity I have some refinment I want to make but it can be a follow up patch, let me know if you have comments on V2 so it will either be part of V3 or a follow up patch

avikivity commented on utils/histogram.hh:
```diff
@@ -318,6 +319,89 @@ public:
+ * summary from that delta histogram.
+ *
+ */
+class summary_calculator {
```
This wants a unit test

amnonh replied:
It will be granted with one

avikivity commented on utils/histogram.hh:
```diff
@@ -318,6 +319,89 @@ public:
+ *
+ */
+class summary_calculator {
+    std::vector<double> _quantiles = { 0.5, 0.95, 0.99};
```
quintiles

avikivity replied:
no, quantile is more accurate

avikivity commented on utils/histogram.hh:
```diff
@@ -318,6 +319,89 @@ public:
+    }
+
+    void set_quantiles(const std::vector<double>& quantiles) noexcept {
+        _quantiles = quantiles;
```
How can this be noexcept?

Should't you reset the other data members?

amnonh replied:
I should probably reset the _summary so it will not return old values, I'll resize it also just in case.

avikivity commented on utils/histogram.hh:
```diff
@@ -318,6 +319,89 @@ public:
+        auto new_entries = _current_histogram.count() - _previous_histogram.count();
+        if (new_entries <= 0) {
+            return;
+        }
```
If the number of entries is lower, surely the result changes.

amnonh replied:
Unless I missunderstand you comment Histogram's count monotonically grow, 
so it's alwasy true that:
``` _current_histogram.count() >= _previous_histogram.count()```

avikivity replied:
Then why do you check for <=?

amnonh replied:
equal should work just as well, I'll change to make it clearer

avikivity commented on utils/histogram.hh:
```diff
@@ -318,6 +319,89 @@ public:
+            total_diff += _current_histogram[i] - _previous_histogram[i];
+            while (pos < _summary.size() && total_diff >= _summary[pos]) {
+                _summary[pos] = _current_histogram.get_bucket_upper_limit(i);
+                pos++;
```
This can overflow.

amnonh replied:
Unless I'm missing your comment, how?
if we are in the loop ```pos < _summary.size()```

avikivity replied:
right

avikivity commented on utils/histogram.hh:
```diff
@@ -318,6 +319,89 @@ public:
+    template <typename Rep, typename Ratio>
+    void mark(std::chrono::duration<Rep, Ratio> dur) noexcept {
+        if (std::chrono::duration_cast<ihistogram::duration_unit>(dur).count() >= 0) {
+            _current_histogram.add(dur);
```
add() is not noexcept.

avikivity commented on api/column_family.cc:
```diff
@@ -79,14 +79,14 @@ future<json::json_return_type>  get_cf_stats(http_context& ctx,
 }
 
```
table.cc need merge?

avikivity replied:
I was referring to the line in the patch changelog that says "table.cc need merge"

amnonh replied:
I figure it out eventualy when going over the commit messages

avikivity commented on service/paxos/paxos_state.cc:
```diff
@@ -225,7 +223,6 @@ future<> paxos_state::learn(storage_proxy& sp, schema_ptr schema, proposal decis
     }).finally([&sp, schema, lc] () mutable {
         auto& stats = sp.get_db().local().find_column_family(schema).get_stats();
         stats.cas_learn.mark(lc.stop().latency());
-        stats.estimated_cas_learn.add(lc.latency());
```
need new unit tests.

amnonh replied:
@avikivity I'm not sure about this comment, do you mean cas needs new unit test?
This is basically the previous functionality (where the the cas had ihistogram an estimated_histogram) only now it uses the unfied ```timed_rate_moving_average_summary_and_histogram``` for both.

But maybe I missunderstood your comment

avikivity replied:
I don't remember why I wrote this here. But we certainly need unit tests to prove the calculations are correct, and a system test (can be manual) that shows the results are reasonable.

avikivity replied:
Ah, it's the very last line in the series. So it's not related to this line but to the series itself.

amnonh replied:
that make sense :) I've added a unit test it will be part of V3

amnonh commented:
@avikivity V3 include the following changes:
1. remove the noexcept where they shluldn't be
2. add a unit test
3. set a longer duration (30s by default) for the summary calculation, the rate is still calculated by 10s intervals but for summaries a longer duration make sense, I set it to 30s so it will not zeroed on low traffic.
4. reset the summary when there are no new entries instead of keeping the last known one.
5. clean the commit message and switch the new entry check from ``` <= 0``` to ``` == 0```
6. pull rebase

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/1540/
